### PR TITLE
feat: better pipeline

### LINKED ISSUE #104: Use pipeline-library to automatically publish docker image
- [x] Apply https://github.com/jenkins-infra/pipeline-library/blob/master/resources/io/jenkins/infra/docker/README.adoc 
- [x] Add repository to infra.ci (either rename it as `docker-account-app` or amend [infra.ci's config](https://github.com/jenkins-infra/charts/blob/main/config/publick8s/jenkins-infra.yaml#L344) )
  - [-] https://github.com/jenkins-infra/kubernetes-management/pull/3031
- ~[ ] Migrate `Jenkinsfile` to `Jenkinsfile_k8s`~ Not neede anymore
- [x] Remove job from trusted.ci.jenkins.io

### PR DESCRIPTION
**Pull request by lemeurherve**
Fixes #104

### CODE REVIEW

File: Jenkinsfile_k8s [hunk 1]
```diff
@@ -6,9 +6,9 @@ podTemplate(
             command: 'cat',
             ttyEnabled: true,
             resourceRequestCpu: '200m',
-            resourceLimitCpu: '200m',
+            resourceLimitCpu: '400m',
             resourceRequestMemory: '256Mi',
-            resourceLimitMemory: '256Mi',
+            resourceLimitMemory: '1024Mi',
```
Comment (dduportal) [inline Jenkinsfile_k8s:11]: - You might want to define the same values for limits and requests (for both CPU and mem)
- Maven is a compiler: don't hesitate to allocate resources "in mass": 2 vCPUs  + 2 Gb memory are quite common for building an app.

Comment (lemeurherve) [reply Jenkinsfile_k8s]: There were "debug" values, it worked with 1024, I'm trying with 512 only, will see if it passes

Comment (lemeurherve) [reply Jenkinsfile_k8s]: 512 wasn't enough, I've put the values you suggested.

File: Dockerfile [hunk 2]
```diff
@@ -1,9 +1,10 @@
 FROM java:openjdk-8-jre-alpine
 
-RUN apk update && apk add --no-cache unzip
-RUN adduser -D -h /home/ircbot -u 1013 ircbot
-ADD target/ircbot-2.0-SNAPSHOT-bin.zip /usr/local/bin/ircbot.zip
-RUN cd /usr/local/bin; unzip ircbot.zip
+RUN apk update && apk add --no-cache unzip=6.0-25ubuntu1 && \
```
Comment (dduportal) [inline Dockerfile:3]:
```suggestion
RUN apk add --no-cache unzip && \
```
When using alpine linux:
- The `apk update` is not needed (it's implied by the flag `--no-cache` of `apk add <...>`)
- You are providing a version of unzip that seems.. Ubuntu-y :D

=> I wonder if unzip is really required: 
- we might want to use the [`ADD`](https://docs.docker.com/engine/reference/builder/#add) instruction that is able to unarchive (but I 'm not sure if zip archives are required)
- And/or we could avoid zipping the jar file during the `make <..>` steps
- And/or using a multi stage image to ensure that `unzip` is not present on the production image

File: Dockerfile [hunk 3]
```diff
@@ -1,9 +1,10 @@
...
+    adduser -D -h /home/ircbot -u 1013 ircbot
+COPY target/ircbot-2.0-SNAPSHOT-bin.zip /usr/local/bin/ircbot.zip
+WORKDIR /usr/local/bin 
+RUN unzip ircbot.zip
```
Comment (dduportal) [inline Dockerfile:7]:
```suggestion
RUN unzip ircbot.zip -d /usr/local/bin
```
Better to keep the default WORKDIR inherited from the parent image (and /usr/local/bin as workdir does not make sense)

Comment (lemeurherve) [reply Dockerfile]: there were a `cd` in this folder before, and nothing to get out of it after, but your suggestion is definitively better.

File: Dockerfile (hunk 2 above)
Comment (lemeurherve) [reply Dockerfile]: Made a jar instead of a zip, and removed unzip :)

**Review by timja — COMMENTED** [review]
err this repo shouldn't be called docker-ircbot, it's a java app, the filter should be adjusted =/

**Comment by lemeurherve** [comment]
> err this repo shouldn't be called docker-ircbot, it's a java app, the filter should be adjusted =/

In my mind, this repository is like all the other tools we're packaging as Docker images for our use in the infra, and thus renaming it felt the right thing to do in order to:
- ease new comers comprehension by having a common pattern between all these tools packaged as docker images
- not having to add a specific entry for each docker packaged repository [in Casc](https://github.com/jenkins-infra/kubernetes-management/blob/40982727e79f0ecc831d068e533630a505a60136/config/jenkins-infra.yaml#L378) (like it was done for "rating")
- facilitate maintenance by aligning all these repositories

@timja sorry if it bothered you, I can revert the renaming if needed.

**Comment by timja** [comment]
If someone asks to make an improvement to the ircbot they should be looking for the ircbot repo not one about docker.
Docker is an implementation detail, if we move it to a function or just running it as part of CI later (which is likely when we move it from Jira to GitHub issues) then it will have nothing to do with docker.

I think we should revert the naming back.

The proper fix would be to not rely on repository naming and instead use topics imo, which would mean deploying the incremental from master of this PR I think:
https://github.com/jenkinsci/github-branch-source-plugin/pull/344

https://repo.jenkins-ci.org/incrementals/org/jenkins-ci/plugins/github-branch-source/2.11.4-rc1561.e708675a334c/github-branch-source-2.11.4-rc1561.e708675a334c.hpi

(for some reason it hasn't been released in 2.5 months)


> ease new comers comprehension by having a common pattern between all these tools packaged as docker images

=/ why does a newcomer need to find them? if on the infra team they can just check the folder on infra.ci.

> facilitate maintenance by aligning all these repositories and applying to them the same kind of pipeline

My only problem with the pipeline bit is that there's no longer any public CI which is not good for an open source project.

**Comment by lemeurherve** [comment]
@timja thanks for these explanations, far more clear now for me!
I didn't knew about the topics, I really like the idea, and prefer it instead of the repo name pattern as it's more explicit and less constraining.
I'll rename the repository, and will look into the topics stuff.
I'll also look into keeping the build on CI (in parallel with the one on infra.ci? Idk yet, will have to check this with @dduportal)

**Comment by dduportal** [comment]
> My only problem with the pipeline bit is that there's no longer any public CI which is not good for an open source project.

Interesting and important point!

Somehow, the "Build + Test + Package as Docker image on a local Docker Engine" should be done on both ci.j and any other instance => "CI".
But here, we also need to "Package as Docker Image + add a GitHub (Release+Tag) to link the Docker image with a tag and change notes + deploy if principal branch or tag) => "CD".

I'm thinking about "how to ensure that we have the same packaging beetween CI and infra.ci". => WDYT about adding a "ifInfraIsTrusted" condition on the docker library for the release/deploy steps?
Since the pipeline libraries are available on all ci instances, the "docker build" could be reused on both.


Then it would only be a setup on the infra.ci's CasC definition (to pick the `Jenkinsfile`) + filtering repositories with topics instead of naming.

**Comment by timja** [comment]
sounds good

**Comment by dduportal** [comment]
Closing as it was superseeded by  #112 and  #113  (at least)

**Pull request closed without merging**